plugins/pipelines/openPr.js

```diff
@@ -0,0 +1,95 @@
+'use strict';
+
+const boom = require('boom');
+const joi = require('joi');
+const schema = require('screwdriver-data-schema');
+const helper = require('./helper');
+const pipelineCheckoutUrlSchema = joi.reach(schema.models.pipeline.create, 'checkoutUrl');
+const pipelineRootDirSchema = joi.reach(schema.models.pipeline.create, 'rootDir');
+
+module.exports = () => ({
+    method: 'POST',
+    path: '/pipelines/openPr',
+    config: {
+        description: 'Open pull request for repository',
+        notes: 'Open pull request',
+        tags: ['api', 'pipelines'],
+        auth: {
+            strategies: ['token'],
+            scope: ['user', '!guest']
+        },
+        plugins: {
+            'hapi-swagger': {
+                security: [{ token: [] }]
+            }
+        },
+        handler: (request, reply) => {
+            const { userFactory } = request.server.app;
+            const { username, scmContext } = request.auth.credentials;
+            const { files, title, message } = request.payload;
+            const checkoutUrl = helper.formatCheckoutUrl(request.payload.checkoutUrl);
+            const rootDir = helper.sanitizeRootDir(request.payload.rootDir);
+
+            return userFactory
+                .get({ username, scmContext })
+                .then(user => {
+                    if (!user) {
+                        throw boom.notFound(`User ${username} does not exist`);
+                    }
+
+                    return user
+                        .unsealToken()
+                        .then(token =>
+                            userFactory.scm.parseUrl({
+                                scmContext,
+                                rootDir,
+                                checkoutUrl,
+                                token
+                            })
+                        )
+                        .then(scmUri => user.getPermissions(scmUri))
+                        .then(permissions => {
+                            if (!permissions.push) {
+                                throw boom.forbidden(`User ${username} does not have push access for this repo`);
+                            }
+                        })
+                        .then(token =>
+                            userFactory.scm.openPr({
+                                checkoutUrl,
+                                files,
+                                token,
+                                scmContext,
+                                title,
+                                message
+                            })
+                        )
+                        .then(pullRequest => {
+                            if (!pullRequest) {
+                                throw boom.notImplemented('openPr not implemented for gitlab');
```

Comment:
so this will always throw an error for scm-gitlab?

Reply:
from what I see, scm-gitlab resolves it with no action https://github.com/screwdriver-cd/scm-gitlab/blob/master/index.js#L888
while scm-base rejects with error https://github.com/screwdriver-cd/scm-base/blob/master/index.js#L584

so this should be okay. (though we didn't need that change in scm-gitlab and let scm-base throw error.)

Reply:
yeah just as Jithin said, this will always throw 501 for scm-gitlab because scm-gitlab returns null.

Reply:
I think I'm confused why opening PR should cause an error when it's not implemented for scm-gitlab
is it a critical operation?

Reply:
that is the correct code https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501
It's up to the caller (UI) to display a nice message when scm-gitlab or bitbucket is where where this feature is not supported. 

It's not a critical feature, UI should handle this case when PR cannot be created and display appropriate message to user.

Reply:
yeah I wasn't confused about the code, just how the experience would be for the user
but if the UI is going to show a better message, then that's fine

Reply:
I think it make sense to return an error, because UI is expecting api to return url when pull request was created. It's up to UI to handle that error properly